### PR TITLE
Add Gero wallet support

### LINKED ISSUE #325: Add Gero wallet support

### PR DESCRIPTION
**Pull request by hariamoor-mlabs**
Closes #325

**Test Plan:** Automated test-cases aren't possible in the absence of Selenium or some other headless-browser. With that said, reviewers are invited to pull the PR and test it locally.

**Details:** Tests were done at development-time using the following [ungoogled] chromium:

```nix
# Nix version: 2.8.0
# NixOS version: 22.05pre375061.c777cdf5c56 (Quokka)
{...}: {
  nixpkgs.config.chromium.enableWideVine = true;

  programs.chromium = {
    package = pkgs.ungoogled-chromium;
    enable = true;
    extensions = [ ... ]; # Extensions enabled for wallets Nami and Gero by Chrome plugin ID
  };
}
```

### CODE REVIEW

File: src/Wallet.purs [hunk 1]
```diff
@@ -167,7 +170,100 @@ dummySign tx@(Transaction { witnessSet: tws@(TransactionWitnessSet ws) }) =
     )
 
 -------------------------------------------------------------------------------
--- FFI stuff
+-- Gero backend
+-------------------------------------------------------------------------------
+-- Mirror of NamiWallet backend for Gero
+type GeroWallet =
```
Comment (ngua) [inline src/Wallet.purs:176]: This is identical to `NamiWallet`. I imagined that the two would be similar, but I didn't foresee that they would be exactly the same. My suggestion would be the following:

- define a type `Cip30Wallet`:
  ```purescript
  type Cip30Wallet (a :: Type) =
    { connection :: Ref.Ref a
    , getWalletAddress :: a -> Aff (Maybe Address)
    , getCollateral :: a -> Aff (Maybe TransactionUnspentOutput)
    , signTx :: a -> Transaction -> Aff (Maybe Transaction)
    }
  ```
- turn `NamiWallet` into `Cpi30Wallet NamiConnection`, similar for `GeroWallet`
- we probably want to de-duplicate `mkNamiWalletAff` and `mkGeroWalletAff`, provided their behavior really is the same (see my other comment below about datums)
- we could also simplify the FFI stuff in this case, e.g. by passing the name of the API object into a set of de-duplicated FFI functions

We can also remove `submitTx` from both wallets, as we now use Ogmios to submit

This would make maintenance a lot easier IMO

File: src/Wallet.purs [hunk 2]
```diff
@@ -167,7 +170,100 @@ dummySign tx@(Transaction { witnessSet: tws@(TransactionWitnessSet ws) }) =
...
+  signTx gero tx = do
+    txHex <- txToHex tx
+    fromGeroHexString (_signTxGero txHex) gero >>= case _ of
+      Nothing -> pure Nothing
+      Just bytes -> map (combineWitnessSet tx) <$> liftEffect
+        ( Deserialization.WitnessSet.convertWitnessSet
+            <$> fromBytesEffect bytes
+        )
+    where
+    -- We have to combine the newly returned witness set with the existing one
```
Comment (ngua) [inline src/Wallet.purs:228]: Does Gero follow Nami in this regard?

File: src/Wallet.purs [hunk 3]
```diff
@@ -195,3 +291,28 @@ foreign import _attachSignature
...
+
+getGeroCollateral :: GeroConnection -> Effect (Promise (Maybe String))
+getGeroCollateral = _getGeroCollateral maybeFfiHelper
+
+foreign import _signTxGero
+  :: String -- Hex-encoded cbor of tx
+  -> GeroConnection
+  -> Effect (Promise String)
+
+foreign import _submitTxGero
```
Comment (ngua) [inline src/Wallet.purs:315]: We can remove this (see my comment above). In fact we should do the same for Nami as well

File: src/Wallet.js [hunk 4]
```diff
@@ -41,6 +41,40 @@ exports._submitTxNami = txHex => nami => () => {
       });
 };
 
+// _enableGero :: Effect (Promise NamiConnection)
+exports._enableGero = () => window.cardano.gero.enable();
```
Comment (ngua) [inline src/Wallet.js:45]: Is this the correct API object? Trying to run `Examples.Gero` gives me:

```
Uncaught TypeError: Cannot read properties of undefined (reading 'enable')
    at exports._enableGero (output.js?ee2e:9826:1)
    at runSync (output.js?ee2e:6816:1)
    at run (output.js?ee2e:7054:1)
    at eval (output.js?ee2e:7365:1)
    at drain (output.js?ee2e:6846:1)
    at Object.enqueue (output.js?ee2e:6867:1)
    at Object.run (output.js?ee2e:7364:1)
    at __do (output.js?ee2e:7970:1)
    at eval (output.js?ee2e:6526:1)
    at Module.eval (output.js?ee2e:6526:1)
exports._enableGero @ output.js?ee2e:9826
runSync @ output.js?ee2e:6816
run @ output.js?ee2e:7054
eval @ output.js?ee2e:7365
drain @ output.js?ee2e:6846
enqueue @ output.js?ee2e:6867
run @ output.js?ee2e:7364
__do @ output.js?ee2e:7970
eval @ output.js?ee2e:6526
eval @ output.js?ee2e:6526
eval @ index.js?2414:8
setTimeout (async)
run @ output.js?ee2e:7254
eval @ output.js?ee2e:7365
drain @ output.js?ee2e:6846
enqueue @ output.js?ee2e:6867
run @ output.js?ee2e:7364
__do @ output.js?ee2e:7970
eval @ output.js?ee2e:6526
eval @ output.js?ee2e:6526
eval @ index.js?2414:8
Promise.then (async)
eval @ index.js?2414:8
./examples/index.js @ bundle.js:17
__webpack_require__ @ bundle.js:656
(anonymous) @ bundle.js:2106
```

File: src/Wallet.js [hunk 5]
```diff
@@ -41,6 +41,40 @@ exports._submitTxNami = txHex => nami => () => {
       });
 };
 
+// _enableGero :: Effect (Promise NamiConnection)
```
Comment (ngua) [inline src/Wallet.js:44]: The type sig in the comment needs to be updated

File: src/Wallet.js [hunk 6]
```diff
@@ -41,6 +41,40 @@ exports._submitTxNami = txHex => nami => () => {
...
+exports._enableGero = () => window.cardano.gero.enable();
+
+// _getGeroAddress :: NamiConnection -> Effect (Promise String)
```
Comment (ngua) [inline src/Wallet.js:47]: Here as well, and for all of the other type signatures in this module

File: src/Wallet.js [hunk 7]
```diff
@@ -41,6 +41,40 @@ exports._submitTxNami = txHex => nami => () => {
...
+exports._signTxGero = txHex => gero => () => {
+  return gero.signTx(txHex, true)
+      .catch(e => {
+          console.log("Error in signTxGero: ", e);
+          throw (JSON.stringify(e));
+      });
+}
+
+// _submitTxGero :: String -> NamiConnection -> Effect (Promise String)
+exports._submitTxGero = txHex => gero => () => {
```
Comment (ngua) [inline src/Wallet.js:70]: As I mentioned in other comments, we can remove this

**Review by ngua — CHANGES_REQUESTED** [review]
The biggest issue at the moment: `_enableGero` throws an exception

File: src/QueryM.purs [hunk 8]
```diff
@@ -1,4 +1,4 @@
--- | TODO docstring
+-- | TODO docstrinn
```
[the remark this answers is not in the document]
Comment (hariamoor-mlabs) [reply src/QueryM.purs:1]: Just a typo, will fix it

File: src/Wallet.purs (hunk 2 above)
Comment (hariamoor-mlabs) [reply src/Wallet.purs:228]: I don't think so

File: src/Wallet.js (hunk 4 above)
Comment (hariamoor-mlabs) [reply src/Wallet.js:45]: Actually, it should be `window.cardano.gerowallet.enable()`

Will fix in the next set of commits

**Comment by ngua** [comment]
Ah, one thing I forgot about in my review yesterday: we probably need to make some changes to the balancer since (I think) it assumes Nami now. @vvtran and I can look at that though

File: examples/AlwaysSucceeds.purs [hunk 9]
```diff
@@ -48,10 +48,12 @@ import Contract.Wallet (mkNamiWalletAff)
 import Data.Argonaut (decodeJson, fromString)
 import Data.BigInt as BigInt
 
+import Effect.Aff (launchAff_)
```
[the remark this answers is not in the document]
Comment (hariamoor-mlabs) [reply examples/AlwaysSucceeds.purs]: Done as of HEAD

File: examples/AlwaysSucceeds.purs [hunk 10]
```diff
@@ -48,10 +48,12 @@ import Contract.Wallet (mkNamiWalletAff)
...
+
 main :: Effect Unit
 main = launchAff_ $ do
-  wallet <- Just <$> mkNamiWalletAff
-  cfg <- over ContractConfig _ { wallet = wallet } <$> traceContractConfig
+  let wallet = mkNamiWalletAff
+  cfg <- traceContractConfig wallet
```
[the remark this answers is not in the document]
Comment (hariamoor-mlabs) [reply examples/AlwaysSucceeds.purs]: Done

File: src/QueryM.purs (hunk 8 above)
[the remark this answers is not in the document]
Comment (hariamoor-mlabs) [reply src/QueryM.purs]: Fixed

File: src/Wallet.js (hunk 4 above)
Comment (hariamoor-mlabs) [reply src/Wallet.js]: Fixed, reopen if any errors persist

File: src/Wallet.js (hunk 6 above)
Comment (hariamoor-mlabs) [reply src/Wallet.js]: Fixed

File: src/Wallet.js (hunk 7 above)
Comment (hariamoor-mlabs) [reply src/Wallet.js]: Fixed

**Comment by klntsky** [comment]
> Ah, one thing I forgot about in my review yesterday: we probably need to make some changes to the balancer since (I think) it assumes Nami now. @vvtran and I can look at that though

I'm going to look into this right now.

File: src/Wallet.purs (hunk 2 above)
Comment (hariamoor-mlabs) [reply src/Wallet.purs:216]: Resolving for now

**Comment by klntsky** [comment]
Now there are conflicts. Could you please cherry-pick [`c559f65` (#455)](https://github.com/Plutonomicon/cardano-transaction-lib/pull/455/commits/c559f656f99a10692534cbd5ecc12f7013056993) @hariamoor-mlabs

**Comment by hariamoor-mlabs** [comment]
@klntsky Still a WIP; could do some more optimizations on the FFI side... Will merge #455 at review-time (ETA: end of week)

File: src/Wallet.purs (hunk 1 above)
Comment (hariamoor-mlabs) [reply src/Wallet.purs]: Removed `NamiConnection` and `GeroConnection` altogether in favor of `Cip30Connection`; we don't need custom logic for submission since we use Ogmios now.

File: examples/AlwaysMints.purs [hunk 11]
```diff
@@ -29,6 +27,10 @@ import Contract.Value as Value
 import Contract.Wallet (mkNamiWalletAff)
 import Data.BigInt as BigInt
 
+import Effect.Aff (launchAff_)
```
Comment (klntsky) [inline examples/AlwaysMints.purs:30]: Please sort imports and remove empty lines.

File: examples/Gero.purs [hunk 12]
```diff
@@ -0,0 +1,33 @@
+-- Examples of how to use the Gero wallet interface through `Contract`
```
Comment (klntsky) [inline examples/Gero.purs:1]: I think this should referring to README instead (duplicate documentation quickly becomes inconsistent). We already have instructions there about `npm run dev`. However, they do not clarify what exactly will be built.

Could you please expand the README describing the process of building examples in more detail?

Also, README mentions Nami as the only wallet supported

> you must use Nami wallet for the time being

This would not be the case after this PR is merged.

File: examples/Gero.purs [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+import Effect.Aff (launchAff_)
+
+main :: Effect Unit
+main = launchAff_ $ do
+  cfg <- defaultContractConfig
+  runContract_ cfg $ do
+    logAction getWalletAddress
+    logAction getWalletCollateral
+
+logAction
```
Comment (klntsky) [inline examples/Gero.purs:28]: I think this function should be inlined since it's trivial and is only used here.

File: src/BalanceTx.purs [hunk 14]
```diff
@@ -387,30 +380,20 @@ balanceTx unattachedTx@(UnattachedUnbalancedTx { unbalancedTx: t }) = do
...
       -- involved with the contract in the unbalanced transaction.
       allUtxos :: Utxo
       allUtxos = utxos `Map.union` utxoIndexToUtxo networkId utxoIndex
 
-      -- After adding collateral, we need to balance the inputs and
-      -- non-Ada outputs before looping, i.e. we need to add input fees
-      -- for the Ada only collateral. No MinUtxos required. In fact perhaps
-      -- this step can be skipped and we can go straight to prebalancer.
-      unbalancedCollTx :: Transaction
-      unbalancedCollTx = addTxCollateral unbalancedTx' collateral
```
Comment (klntsky) [inline src/BalanceTx.purs:406]: This function should NOT be removed.

`addTxCollateral` is still needed for `Nami`.

Please use the same logic as in https://github.com/Plutonomicon/cardano-transaction-lib/pull/455/commits/c559f656f99a10692534cbd5ecc12f7013056993

File: src/QueryM/Utxos.purs [hunk 15]
```diff
@@ -41,68 +38,41 @@ import QueryM.Ogmios as Ogmios
...
+      out
+        :: Maybe
+             ( Array
+                 ( TransactionInput /\
+                     TransactionOutput
+                 )
+             )
+      out = out' <#> bisequence # sequence
+    in
+      (wrap <<< Map.fromFoldable) <$> out
```
Comment (klntsky) [inline src/QueryM/Utxos.purs:71]:
```suggestion
      wrap <<< Map.fromFoldable <$> out
```

File: src/Wallet.js [hunk 16]
```diff
@@ -7,40 +7,34 @@ if (typeof BROWSER_RUNTIME != 'undefined' && BROWSER_RUNTIME) {
...
 
-// _enableNami :: Effect (Promise NamiConnection)
-exports._enableNami = () => window.cardano.nami.enable();
+ALLOWED_APIS = new Set([ "nami", "gerowallet" ]);
 
-// _getNamiAddress :: NamiConnection -> Effect (Promise String)
-exports._getNamiAddress = nami => () =>
+// _enable :: String -> Effect (Promise Cip30Connection)Q
+exports._enable = api => () =>
+  ALLOWED_API.has(api) ? eval(`window.cardano.${api}.enable()`) : undefined;
```
Comment (klntsky) [inline src/Wallet.js:14]: `eval` call is definitely not needed here.

Use `cardano[api].enable()`.

Also no need to use ternary operator:

```
exports._enable = api => () => {
  if (ALLOWED_API.has(api)) {
    window.cardano[api].enable()
  }
}
```

**Review by klntsky — CHANGES_REQUESTED** [review]
Thank you very much for this work! A few changes are still needed here and there

File: src/BalanceTx.purs (hunk 14 above)
Comment (hariamoor-mlabs) [reply src/BalanceTx.purs:406]: Unfortunately, that code won't work, because I got rid of the `Nami _` case for `Wallet` type; it's all a newtype to `Cip30Wallet` now. We also don't use Nami for submission, preferring Ogmios, AFAIK. Could you give me further details about the logic to implement in the merge?

File: src/Wallet.purs [hunk 17]
```diff
@@ -174,30 +100,110 @@ dummySign tx@(Transaction { witnessSet: tws@(TransactionWitnessSet ws) }) =
...
+    , getCollateral
+    , signTx
+    , signTxBytes
+    }
+
+-------------------------------------------------------------------------------
+-- Helper functions
+-------------------------------------------------------------------------------
+
+enable :: String -> Aff Cip30Connection
```
Comment (ngua) [inline src/Wallet.purs:126]: I would prefer to avoid a stringly-typed identifier for different wallets, since there are really only two options we can support at the moment

File: src/Contract/Wallet.purs [hunk 18]
```diff
@@ -1,17 +1,19 @@
...
 
 import Contract.Monad (Contract)
 import Effect.Aff.Class (liftAff)
 import Wallet
-  ( NamiConnection
-  , NamiWallet
-  , Wallet(Nami)
+  ( Cip30Connection
+  , Cip30Wallet
+  , Wallet(..)
```
Comment (ngua) [inline src/Contract/Wallet.purs:14]: Explicit imports please

File: examples/Nami.purs [hunk 19]
```diff
@@ -13,10 +13,11 @@ import Contract.Address (getWalletAddress, getWalletCollateral)
 import Contract.Monad
   ( Contract
   , defaultContractConfig
-  , launchAff_
```
Comment (ngua) [inline examples/Nami.purs:16]: Why remove this? It's re-exported from `Contract.Monad` so users don't have to add `aff` to their dependencies in spago.dhall

File: examples/Gero.purs [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+
+import Contract.Prelude
+import Contract.Address (getWalletAddress, getWalletCollateral)
+import Contract.Monad
+  ( Contract
+  , defaultContractConfig
+  , runContract_
+  )
+
+import Effect.Aff (launchAff_)
```
Comment (ngua) [inline examples/Gero.purs:19]: See my other comment about this

File: src/QueryM.purs [hunk 21]
```diff
@@ -165,10 +164,13 @@ import Types.Natural (Natural)
 import Types.PlutusData (PlutusData)
 import Types.PubKeyHash (PaymentPubKeyHash, PubKeyHash, StakePubKeyHash)
 import Types.Scripts (PlutusScript)
-import Types.Transaction (TransactionHash)
 import Types.UsedTxOuts (newUsedTxOuts, UsedTxOuts)
 import Untagged.Union (asOneOf)
-import Wallet (Wallet(Nami), NamiWallet, NamiConnection)
+import Wallet
+  ( Wallet(..)
```
Comment (ngua) [inline src/QueryM.purs:170]: Explicit imports

File: src/Wallet.purs (hunk 17 above)
Comment (ngua) [reply src/Wallet.purs:126]: We could have `data WalletApi = Nami | Gero` or something and just convert it to a string

File: src/Wallet.purs [hunk 22]
```diff
@@ -35,121 +36,46 @@ import Helpers ((<<>>))
...
--- Nami backend
--------------------------------------------------------------------------------
--- Record-of-functions for real or mocked Nami wallet, includes `Ref` to
--- connection (e.g. with `window.cardano.nami` as a `NamiConnection`)
-type NamiWallet =
-  { -- A reference to a connection with Nami, i.e. `window.cardano.nami`
-    connection :: Ref.Ref NamiConnection
+type Cip30Wallet =
+  { -- A reference to a connection with the wallet, i.e. `window.cardano.nami`
+    connection :: Ref.Ref Cip30Connection
```
Comment (ngua) [inline src/Wallet.purs:45]: Since this PR totally reworks the `Wallet` type, we can actually remove the `Ref` and hold the connection directly. The original idea was to have a ref to the API object since it will mutate if you call `enable` on it. But we actually call `enable` directly during initialization and then use the resulting API object, so we never need to `write` the ref. We can just totally eliminate the overhead of having a `Ref` hold it

Apologies for not pointing this out earlier

File: src/Wallet.js [hunk 23]
```diff
@@ -7,40 +7,34 @@ if (typeof BROWSER_RUNTIME != 'undefined' && BROWSER_RUNTIME) {
...
+
+// _getAddress :: Cip30Connection -> Effect (Promise String)
+exports._getAddress = nami => () =>
```
Comment (ngua) [inline src/Wallet.js:17]: We should rename all of the `nami` arguments to `api` or something

File: src/QueryM/Utxos.purs (hunk 15 above)
Comment (ngua) [reply src/QueryM/Utxos.purs:71]: Let's make sure that we can actually remove Nami-specific filtering of utxos. I know the comment said that it would be possible after switching to Omgios for submission, but I don't know if that's actually true (in fact I suspect it might not be). We might also end up needing to filter out the collateral for Gero as well

File: src/Wallet.js (hunk 16 above)
Comment (vvtran) [reply src/Wallet.js:14]: Sorry for the potentially silly question but does this need to be `ALLOWED_APIS` (note plural)

File: src/BalanceTx.purs (hunk 14 above)
Comment (vvtran) [reply src/BalanceTx.purs:406]: We do need to add this back in, especially if Gero have their own add collateral functionality.

File: src/Wallet.purs [hunk 24]
```diff
@@ -174,30 +100,110 @@ dummySign tx@(Transaction { witnessSet: tws@(TransactionWitnessSet ws) }) =
...
+enable = Promise.toAffE <<< _enable
+
+txToHex :: Transaction -> Aff String
+txToHex =
+  liftEffect
+    <<< map (byteArrayToHex <<< Serialization.toBytes <<< asOneOf)
+    <<< Serialization.convertTransaction
+
+getWalletAddress :: Cip30Connection -> Aff (Maybe Address)
+getWalletAddress nami = fromHexString _getAddress nami <#>
```
Comment (vvtran) [inline src/Wallet.purs:136]: let's change `nami` argument.

File: src/Wallet.purs [hunk 25]
```diff
@@ -174,30 +100,110 @@ dummySign tx@(Transaction { witnessSet: tws@(TransactionWitnessSet ws) }) =
...
+      )
+  where
+  -- We have to combine the newly returned witness set with the existing one
+  -- Otherwise, any datums, etc... won't be retained
+  combineWitnessSet :: Transaction -> TransactionWitnessSet -> Transaction
+  combineWitnessSet (Transaction tx'@{ witnessSet: oldWits }) newWits =
+    Transaction $ tx' { witnessSet = oldWits <> newWits }
+
+signTxBytes :: Cip30Connection -> CborBytes -> Aff (Maybe CborBytes)
+signTxBytes nami txBytes = do
```
Comment (vvtran) [inline src/Wallet.purs:165]: `nami`

**Review by vvtran — COMMENTED** [review]
Comments. Mainly, we do need the Gero collateral added as @ngua  says.

**Comment by klntsky** [comment]
I will make a PR to this branch soon, resolving all problems from review above.

**Comment by klntsky** [comment]
I [applied some code suggestions](https://github.com/Plutonomicon/cardano-transaction-lib/pull/474), however the PR is still not ready: AlwaysMints does not work with Gero for some reason. Investigating.

File: examples/Gero.purs [hunk 26]
```diff
@@ -0,0 +1,33 @@
...
+
+main :: Effect Unit
+main = launchAff_ $ do
+  cfg <- defaultContractConfig
```
Comment (klntsky) [inline examples/Gero.purs:23]: It does not use Gero wallet

Comment (vvtran) [reply examples/Gero.purs:23]: Good catch.

File: src/QueryM.purs [hunk 27]
```diff
@@ -337,15 +338,12 @@ withMWalletAff
   :: forall (a :: Type). (Wallet -> Aff (Maybe a)) -> QueryM (Maybe a)
 withMWalletAff act = asks _.wallet >>= maybe (pure Nothing) (liftAff <<< act)
 
-callNami
+callWallet
```
Comment (ngua) [inline src/QueryM.purs:341]: Maybe this should be `callCip30Wallet`, since it won't work with the testing wallet for example

File: src/QueryM/Utxos.purs (hunk 15 above)
Comment (ngua) [reply src/QueryM/Utxos.purs:71]: In fact, considering Nami's collateral mechanism has been added to CIP-30, we might need to filter out the collateral utxos for Gero as well

**Comment by klntsky** [comment]
I still need to find what to do with collateral filtering for both Nami and Gero.

**Comment by ngua** [comment]
We have a lot of critical issues at the moment, so I'm going to convert this back to a draft

File: examples/Datums.purs [hunk 28]
```diff
@@ -19,11 +19,13 @@ module Examples.Datums (main) where
 
 import Contract.Prelude
 
-import Contract.Monad (runContract_, launchAff_, logInfo', traceContractConfig)
+import Contract.Monad (runContract_, logInfo', traceContractConfig)
 import Contract.PlutusData (DataHash, getDatumByHash, getDatumsByHashes)
 import Contract.Prim.ByteArray (hexToByteArrayUnsafe)
 import Data.Newtype (wrap)
 
+import Effect.Aff (launchAff_)
```
Comment (ngua) [inline examples/Datums.purs:27]: I'd like to revert this change, not sure why this file was even modified

File: examples/AlwaysSucceeds.purs [hunk 29]
```diff
@@ -8,19 +8,18 @@ import Contract.Prelude
 import Contract.Address (scriptHashAddress)
 import Contract.Aeson (decodeAeson, fromString)
 import Contract.Monad
-  ( ContractConfig(ContractConfig)
-  , launchAff_
+  ( Contract
+  , ContractConfig(..)
```
Comment (ngua) [inline examples/AlwaysSucceeds.purs:12]: Explicit imports

File: examples/AlwaysMints.purs (hunk 11 above)
Comment (ngua) [reply examples/AlwaysMints.purs:29]: Still needs to be done

File: examples/AlwaysMints.purs [hunk 30]
```diff
@@ -28,6 +26,8 @@ import Contract.TxConstraints as Constraints
 import Contract.Value as Value
 import Contract.Wallet (mkNamiWalletAff)
 import Data.BigInt as BigInt
+import Effect.Aff (launchAff_)
+import Types.Scripts (MintingPolicy)
```
Comment (ngua) [inline examples/AlwaysMints.purs:30]: We should always import from `Contract.*` in the examples

File: examples/Gero.purs [hunk 31]
```diff
@@ -0,0 +1,21 @@
...
+-- Make sure that you have the Gero browser extension installed. Allow the page
+-- to access your wallet when prompted by Gero
+--
+-- NOTE: Gero has the same limitations as Nami, i.e., Chromium-based only
+module Examples.Gero (main) where
+
+import Contract.Prelude
+
+import Contract.Address (getWalletAddress, getWalletCollateral)
+import Contract.Monad (ContractConfig(..), runContract_, traceContractConfig)
```
Comment (ngua) [inline examples/Gero.purs:11]: Explicit imports

File: examples/AlwaysSucceeds.purs [hunk 32]
```diff
@@ -29,18 +28,18 @@ import Contract.Transaction
...
-import Types.Scripts (ValidatorHash)
+import Types.Scripts (Validator, ValidatorHash)
 import Types.Transaction (TransactionInput(TransactionInput), TransactionHash)
 import Types.TxConstraints (TxConstraints)
+import Wallet (mkGeroWalletAff)
 
 main :: Effect Unit
 main = launchAff_ $ do
-  wallet <- Just <$> mkNamiWalletAff
+  wallet <- Just <$> mkGeroWalletAff
```
Comment (ngua) [inline examples/AlwaysSucceeds.purs:42]: I'd like to keep all examples using Nami, since users are more likely to have that installed

File: src/Contract/Wallet.purs [hunk 33]
```diff
@@ -1,17 +1,19 @@
...
   , module Wallet
   ) where
 
 import Contract.Monad (Contract)
 import Effect.Aff.Class (liftAff)
 import Wallet
-  ( NamiConnection
-  , NamiWallet
-  , Wallet(Nami)
+  ( Cip30Connection
```
Comment (ngua) [inline src/Contract/Wallet.purs:12]: We are we exporting things like `Cip30Connection`? That should be considered internal

File: src/QueryM/Utxos.purs [hunk 34]
```diff
@@ -84,17 +82,18 @@ utxosAt addr = asks _.wallet >>= maybe (allUtxosAt addr) (utxosAtByWallet addr)
...
       in
-        (wrap <<< Map.fromFoldable) <$> out
+        wrap <<< Map.fromFoldable <$> out
 
-  -- Nami appear to remove collateral from the utxo set, so we shall do the same.
-  -- This is crucial if we are submitting via Nami. If we decide to submit with
-  -- Ogmios, we can remove this.
+  -- cip30 wallets appear to remove collateral from the utxo set, so we shall do
+  -- the same.
+  -- This is crucial if we are submitting via Nami/Gero. If we decide to submit
```
Comment (ngua) [inline src/QueryM/Utxos.purs:89]: I think this comment should be removed, as we don't submit via wallets anymore

File: src/Wallet.purs [hunk 35]
```diff
@@ -174,30 +99,107 @@ dummySign tx@(Transaction { witnessSet: tws@(TransactionWitnessSet ws) }) =
           "ed25519_sig1ynufn5umzl746ekpjtzt2rf58ep0wg6mxpgyezh8vx0e8jpgm3kuu3tgm453wlz4rq5yjtth0fnj0ltxctaue0dgc2hwmysr9jvhjzswt86uk"
     )
 
+-------------------------------------------------------------------------------
+-- Gero backend
+-------------------------------------------------------------------------------
+
+mkGeroWalletAff :: Aff Wallet
```
Comment (ngua) [inline src/Wallet.purs:106]: We can factor this out (along with `mkNamiWalletAff`) into a `mkCip30WalletAff` that takes an `Effect (Promise Cip30Connection)` HOF argument, along with the name of the wallet as a `String` to include in the error message. It will make it easier to add more CIP-30-compliant wallets in the future

File: src/BalanceTx.purs [hunk 36]
```diff
@@ -530,6 +526,12 @@ balanceTx unattachedTx@(UnattachedUnbalancedTx { unbalancedTx: t }) = do
   feeBuffer :: BigInt
   feeBuffer = fromInt 500000
 
+-- Nami provides a 5 Ada collateral that we should add the tx before balancing.
```
Comment (ngua) [inline src/BalanceTx.purs:529]: This comment should be updated to reflect the fact that all CIP-30 wallets do this now

File: src/Wallet.purs [hunk 37]
```diff
@@ -29,127 +30,51 @@ import Effect (Effect)
...
   -- multiple addresses)
-  , getWalletAddress :: NamiConnection -> Aff (Maybe Address)
+  , getWalletAddress :: Cip30Connection -> Aff (Maybe Address)
   -- Get the collateral UTxO associated with the Nami wallet
-  , getCollateral :: NamiConnection -> Aff (Maybe TransactionUnspentOutput)
+  , getCollateral :: Cip30Connection -> Aff (Maybe TransactionUnspentOutput)
   -- Sign a transaction with the current wallet
-  , signTx :: NamiConnection -> Transaction -> Aff (Maybe Transaction)
+  , signTx :: Cip30Connection -> Transaction -> Aff (Maybe Transaction)
   -- Sign a transaction with the current wallet
```
Comment (ngua) [inline src/Wallet.purs:52]: Not related to the current changes but this comment should be made more accurate

**Review by ngua — CHANGES_REQUESTED** [review]
We also need to update the readme and comments that say only Nami is supported

File: src/Contract/Wallet.purs (hunk 33 above)
Comment (klntsky) [reply src/Contract/Wallet.purs:12]:
```
  An export for Cip30Wallet requires the following to also be exported:

    Cip30Connection
```

File: examples/AlwaysMints.purs [hunk 38]
```diff
@@ -28,6 +27,7 @@ import Contract.TxConstraints as Constraints
 import Contract.Value as Value
 import Contract.Wallet (mkNamiWalletAff)
 import Data.BigInt as BigInt
+import Contract.Scripts (MintingPolicy)
```
Comment (ngua) [inline examples/AlwaysMints.purs:30]: I'm not sure why imports in unrelated modules have gotten changed so much on this branch

File: src/QueryM/Utxos.purs [hunk 39]
```diff
@@ -26,12 +23,12 @@ import Effect.Class (liftEffect)
...
 import QueryM (QueryM, getWalletCollateral, mkOgmiosRequest)
+import QueryM.Ogmios as Ogmios
 import Serialization.Address (Address)
-import Types.Transaction (TransactionInput)
 import TxOutput (ogmiosTxOutToTransactionOutput, txOutRefToTransactionInput)
+import Types.Transaction (TransactionInput)
 import Types.UsedTxOuts (UsedTxOuts, isTxOutRefUsed)
-import Wallet (Wallet(Nami))
-import QueryM.Ogmios as Ogmios
+import Wallet (Wallet(..))
```
Comment (ngua) [inline src/QueryM/Utxos.purs:31]: Explicit imports

**Review by ngua — APPROVED** [review]
Just two small things below, then we can merge this
